errorshow: simplify printing of keyword argument types using a new macro format

aviatesk opened the pull request:
In Julia, keyword arguments are represented as `Base.Pairs` objects. However, the object type often appears unnecessarily complex, especially when printed in a stack trace.

This commit aims to simplify the printing of stack traces that involve keyword method calls, while still allowing us to reconstruct the actual method signature types from the printed signature types.

The approach is similar to #49117: this commit introduces a new macro called `Base.@Kwargs`. It follows the same syntax as `@NamedTuple` and returns a `Base.Pairs` type that is used for keyword method calls. We use this syntax when printing keyword argument types.

Here's an example of a stack trace for `sum("julia"; init=1)`:
```diff
diff --git a/b.jl b/a.jl
index 91dd6f0464..b804ae4be5 100644
--- a/b.jl
+++ b/a.jl
@@ -22,12 +22,11 @@ Stacktrace:
     @ Base ./reduce.jl:44 [inlined]
   [6] mapfoldl(f::typeof(identity), op::typeof(Base.add_sum), itr::String; init::Int64)
     @ Base ./reduce.jl:175 [inlined]
-  [7] mapreduce(f::typeof(identity), op::typeof(Base.add_sum), itr::String; kw::Base.Pairs{…})
+  [7] mapreduce(f::typeof(identity), op::typeof(Base.add_sum), itr::String; kw::@Kwargs{init::Int64})
     @ Base ./reduce.jl:307 [inlined]
-  [8] sum(f::typeof(identity), a::String; kw::Base.Pairs{Symbol, Int64, Tuple{Symbol}, @NamedTuple{init::Int64}})
+  [8] sum(f::typeof(identity), a::String; kw::@Kwargs{init::Int64})
     @ Base ./reduce.jl:535 [inlined]
-  [9] sum(a::String; kw::Base.Pairs{Symbol, Int64, Tuple{Symbol}, @NamedTuple{init::Int64}})
+  [9] sum(a::String; kw::@Kwargs{init::Int64})
     @ Base ./reduce.jl:564 [inlined]
  [10] top-level scope
```

Feel free to share any comments or suggestions. If this idea seems acceptable, I will add test cases and also address any broken test cases.

aviatesk committed:
RFC: errorshow: simplify printing of keyword argument types using a new macro format

In Julia, keyword arguments are represented as `Base.Pairs` objects.
However, the object type often appears unnecessarily complex,
especially when printed in a stack trace.

This commit aims to simplify the printing of stack traces that involve
keyword method calls, while still allowing us to reconstruct the actual
method signature types from the printed signature types.

The approach is similar to #49117: this commit introduces a new macro
called `Base.@Kwargs`. It follows the same syntax as `@NamedTuple` and
returns a `Base.Pairs` type that is used for keyword method calls.
We use this syntax when printing keyword argument types.

Here's an example of a stack trace:
```diff
diff --git a/b.jl b/a.jl
index 91dd6f0464..b804ae4be5 100644
--- a/b.jl
+++ b/a.jl
@@ -22,12 +22,11 @@ Stacktrace:
     @ Base ./reduce.jl:44 [inlined]
   [6] mapfoldl(f::typeof(identity), op::typeof(Base.add_sum), itr::String; init::Int64)
     @ Base ./reduce.jl:175 [inlined]
-  [7] mapreduce(f::typeof(identity), op::typeof(Base.add_sum), itr::String; kw::Base.Pairs{…})
+  [7] mapreduce(f::typeof(identity), op::typeof(Base.add_sum), itr::String; kw::Base.@Kwargs{init::Int64})
     @ Base ./reduce.jl:307 [inlined]
-  [8] sum(f::typeof(identity), a::String; kw::Base.Pairs{Symbol, Int64, Tuple{Symbol}, @NamedTuple{init::Int64}})
+  [8] sum(f::typeof(identity), a::String; kw::Base.@Kwargs{init::Int64})
     @ Base ./reduce.jl:535 [inlined]
-  [9] sum(a::String; kw::Base.Pairs{Symbol, Int64, Tuple{Symbol}, @NamedTuple{init::Int64}})
+  [9] sum(a::String; kw::Base.@Kwargs{init::Int64})
     @ Base ./reduce.jl:564 [inlined]
  [10] top-level scope
```

Feel free to share any comments or suggestions. If this idea seems
acceptable, I will add test cases and also address any broken test cases.

KristofferC commented:
If it is exported then the `Base.` prefix in `Base.@Kwargs{init::Int64}` would disappear. That seems worth it to me.

aviatesk commented:
I didn't export it because I considered it as an internal implementation detail that keyword arguments are represented as `Base.Pairs`. But I'm fine with exporting it as I agree with the benefit of further simplifying the stacktrace.

KristofferC commented:
> I didn't export it because I considered it as an internal implementation detail that keyword arguments are represented as Base.Pairs.

If that implementation changes then `@Kwargs` would also change. I don't think a user of `@Kwargs` can expect that the return type is fixed, only that it returns the same one as used for keyword arguments.

aviatesk committed:
export and document `Base.@Kwargs` and further simplify the stack trace view

aviatesk committed:
use the `@Kwargs` syntax only when printing kwmethod signature within stack trace view

And add tests.

aviatesk commented on base/show.jl:
```diff
@@ -1094,30 +1111,41 @@ function show_datatype(io::IO, x::DataType, wheres::Vector{TypeVar}=TypeVar[])
             print(io, "}")
             return
         end
+    elseif get(io, :simplify_kwstype, false) && kwsnt !== nothing
```
I enabled the printing with the `@Kwargs`-macro syntax only within stack trace view. Please let me know if you think it is more reasonable to enable the printing in the other situations.

The main reasoning behind the current choice is that `Base.Pairs` is a general data structure and its usage is not restricted to represent keyword arguments.

aviatesk committed:
add news entry

aviatesk committed:
more type stability

aviatesk commented on test/errorshow.jl:
```diff
@@ -995,6 +995,20 @@ bt_str = sprint(Base.show_backtrace, bt)
+    end
+    @test res
+    bt_str = sprint(Base.show_backtrace, bt)
+    @test occursin("simplify_kwargs_type(pos::Int64; kws::@Kwargs{kw1::Float64, kw2::String})", bt_str)
```
```suggestion
    @test occursin("simplify_kwargs_type(pos::$Int; kws::@Kwargs{kw1::Float64, kw2::String})", bt_str)
```

aviatesk commented on test/show.jl:
```diff
@@ -1366,6 +1366,8 @@ test_repr("(:).a")
 @test repr(@NamedTuple{a::Float64, b}) == "@NamedTuple{a::Float64, b}"
 
+# Test general printing of `Base.Pairs` (it should not use the `@Kwargs` macro syntax)
+@test repr(@Kwargs{init::Int}) == "Base.Pairs{Symbol, Int64, Tuple{Symbol}, @NamedTuple{init::Int64}}"
```
```suggestion
@test repr(@Kwargs{init::Int}) == "Base.Pairs{Symbol, $Int, Tuple{Symbol}, @NamedTuple{init::$Int}}"
```

aviatesk committed:
Apply suggestions from code review

aviatesk committed:
enable the type-repr simplification unconditionally in the stack trace

Since keyword pairs can appear within positional arguments, it can be
confusing if we print the same type with different representations.

aviatesk committed:
omit type annotation for splat keyword argument

aviatesk commented:
I'd argue that 6 👍 means this PR is preferable. I'm planning to merge this PR if there are no objections.

aviatesk committed:
add test for `@Kwargs`

aviatesk committed:
clean up test/errorshow.jl
